Switch bindings implementation to `nanobind`

nicholasjng opened the pull request:
Title is self-explanatory. This uses `nanobind` to build Python bindings for Google Benchmark instead of `pybind11`, with the following consequences:

- Python 3.7 support is dropped, since nanobind needs PEP590 vectorcalls, first added in Python 3.8.
- `nanobind` is checked in as a bazel dep, currently at HEAD instead of a tag, because some features used to generate the GBM binding code have not made it into a release yet.
- The minimum MacOS version was bumped up to 10.14 (Mojave), to get full C++17 support.

For performance considerations and changes in contrast to the current implementation, see the discussion #1521.

(For reviewers: I am not sure of the options I gave in the nanobind BUILD file - I looked at what compile commands CMake generated in release mode for nanobind and went on from there.)

The bindings now also contain the state name from #1511.

dmah42 commented on bindings/python/nanobind.BUILD:
```diff
@@ -0,0 +1,41 @@
+        "-Os",  # size optimization option for nanobind
+        "-flto", # enable LTO
+    ],
+    linkopts = ["-undefined suppress", "-flat_namespace", "-flto"],
```
seems like -flat_namespace is not recognised by gcc (see test failure)

nicholasjng commented:
Hmm, I guess I need to pass the `copts` to nanobind via a `select` as well. The `cxxopt` for C++17 is different as well on Windows, I will get to it asap.

nicholasjng commented:
I fixed the MSVC breakage, which was due to (my) user error. Should be good to go now, binding test matrix is all green, other failures are due a downed apt mirror presumably.

If you want, I can restructure the PR and split it up (e.g. carving out the CI parts), to make it easier to revert to `pybind11` in the event, without touching the other changes here. Let me know what you think.
